llama.cpp

```diff
@@ -13085,6 +13085,25 @@ static std::vector<llama_grammar_candidate> llama_grammar_reject_candidates(
 // grammar - external
 //
 
+enum detect_left_recursion_status {
+    // haven't searched this nonterminal
+    LLAMA_LEFT_REC_NOT_SEARCHED    = 0,
+
+    // searching this nonterminal in progress
+    LLAMA_LEFT_REC_IN_PROGRESS     = 1,
+
+    // finished searching this nonterminal
+    LLAMA_LEFT_REC_FINISHED_SEARCH = 2,
+
+    // detected a cycle
+    LLAMA_LEFT_REC_FOUND_CYCLE     = 3,
+};
+
+static void detect_left_recursion(
```

Comment:
Minor formatting note -- internal functions and structures such as these should go in the section marked `grammar - internal`.

Also, because this is a grammar-specific function, it should probably be renamed `llama_grammar_detect_left_recursion`